[WIP] update env var doc (#4186)
circleci/circleci-docs: jekyll/_cci2/env-vars.md

## Overview
{:.no_toc}

There are several ways to use environment variables in CircleCI to provide variety in scope and authorization level. Environment variables are governed by an [order of precedence](#order-of-precedence), depending on how they are set, allowing control at each level in your configuration. 

To add **private keys** or **secret environment variables** for use throughout your private project, use the [Environment Variables page under Project Settings](#setting-an-environment-variable-in-a-project) in the CircleCI application. The variable values are neither readable nor editable in the app after they are set. To change the value of an environment variable, delete the current variable and add it again with the new value.

> Private environment variables enable you to store secrets safely even when your project is public, see [Building Open Source Projects]({{ site.baseurl }}/2.0/oss/) for associated settings information. 

Reviewer: ```suggestion
Private environment variables enable you to store secrets safely even when your project is public. Refer to the [Building Open Source Projects]({{ site.baseurl }}/2.0/oss/) page for associated settings information. 
```

> Use Contexts to [further restrict access to environment variables](#setting-an-environment-variable-in-a-context). Contexts are set from the Organization Settings in the CircleCI application. For more information refer to the [Restricting a Context]({{ site.baseurl }}/2.0/contexts/#restricting-a-context) documentation.

Reviewer: ```suggestion
Use Contexts to [further restrict access to environment variables](#setting-an-environment-variable-in-a-context). Contexts are set from the Organization Settings in the CircleCI application. For more information about setting setting contexts, refer to the [Restricting a Context]({{ site.baseurl }}/2.0/contexts/#restricting-a-context) documentation.
```

Author: I think I'll change this to "for more information about controlling access to env vars using contexts. ..."

> Use Contexts to [further restrict access to environment variables](#setting-an-environment-variable-in-a-context). Contexts are set from the Organization Settings in the CircleCI application. For more information refer to the [Restricting a Context]({{ site.baseurl }}/2.0/contexts/#restricting-a-context) documentation.

Author: ```suggestion
Use Contexts to [further restrict access to environment variables](#setting-an-environment-variable-in-a-context). Contexts are set from the Organization Settings in the CircleCI application. For more information about controlling access to env vars with Contexts, refer to the [Restricting a Context]({{ site.baseurl }}/2.0/contexts/#restricting-a-context) documentation.
```

      - run: # print the name of the branch we're on
          name: "What branch am I on?"
          command: echo ${CIRCLE_BRANCH}
      # Run another step, the same as above; note that you can
      # invoke environment variable without curly braces.
      - run:
          name: "What branch am I on now?"
          command: echo $CIRCLE_BRANCH
      - run:
          name: "What was my custom environment variable?"
          command: echo ${MY_ENV_VAR}
      - run:
          name: "Print an env var stored in the Project"
          command: echo ${PROJECT_ENV_VAR}
      - run:
          name: "Print an env var stored in a Context"
          command: echo ${CONTEXT_ENV_VAR}

workflows: # a single workflow with a single job called build
  build:
    jobs:
      - build:
          context: Testing-Env-Vars
```

The above `config.yml` demonstrates the following: 

- Setting custom environment variables
- Reading a built-in environment variable that CircleCI provides (`CIRCLE_BRANCH`)
- How variables are used (or interpolated) in your `config.yml`
- Secrets masking, applied to environment variable set in the project or within a Context.

> When the above config runs, the output looks like this, notice the env var stored in the Project is masked, and displays as `****`:

Reviewer: ```suggestion
When the above config runs, the output looks like this. Notice the env var stored in the Project is masked, and displays as `****`:
```

working_directory: /go/src/github.com/$ORGNAME/$REPONAME
```

Using `version:2.1` config, you can reuse pieces of config across your  `config.yml`. By using the `parameters` declaration, you can interpolate (or, "pass values") into reusable `commands` `jobs` and `executors`:

```yaml
version: 2.1

jobs:
  parameters:
    org_name:
      type: string
      default: my_org
    repo_name:
      type: string
      default: my_repo
  my_job:
    docker:
      - circleci/node:12.15.0
    working_directory: /go/src/github.com/<< parameters.org_name >>/<< parameters.repo_name >>
```

> Consider reading the documentation on [using the parameters declaration]({{ site.baseurl }}/2.0/reusing-config/#using-the-parameters-declaration).

Reviewer: ```suggestion
For more information, read the documentation on [using the parameters declaration]({{ site.baseurl }}/2.0/reusing-config/#using-the-parameters-declaration).
```

The `$BASH_ENV` workaround only works with `bash`. Other shells probably won't work.

### Alpine Linux
{:.no_toc}

An image that's based on [Alpine Linux](https://alpinelinux.org/) (like [docker](https://hub.docker.com/_/docker)), uses the `ash` shell. 

To use environment variables with `ash`, just add these 2 parameters to your job.

```yaml
version: 2.1

jobs:
  build:    
    shell: /bin/sh -leo pipefail
    environment:
      - BASH_ENV: /etc/profile
```

## Setting an Environment Variable in a Shell Command

> While CircleCI does not support interpolation when setting environment variables, it is possible to set variables for the current shell by [using `BASH_ENV`](#using-parameters-and-bashenv). This is useful for both modifying your `PATH` and setting environment variables that reference other variables.

Reviewer 2: ```suggestion
While CircleCI does not support interpolation when setting environment variables, it is possible to set variables for the current shell by [using `BASH_ENV`](#using-parameters-and-bash_env). This is useful for both modifying your `PATH` and setting environment variables that reference other variables.
```

Author: ohhhh right! Thanks so much for this @teesloane it's been doing my head in. Will see if it works :-)

## Setting an Environment Variable in a Context

1. In the CircleCI application, go to your organization settings by clicking the link in the left hand navigation.

    ![Contexts]({{ site.baseurl }}/assets/img/docs/org-settings-contexts-v2.png)

2. Select the Context you want to associate your envorinment variable with, or create a new one by clicking the Create Context button.
3. Click Add Environment Variable and enter a name and value.
4. Use your new environment variable in your `.circleci/config.yml` once the context is added under the workflows key, as follows:

```yaml
version: 2.1

workflows:
  test-env-vars:
    jobs:
      - build:
          context: my_context_name # has an env var called MY_ENV_VAR

jobs:
  build:
    docker:
      - image: cimg/base:2020.01
    steps:
      - checkout
      - run: 
          name: "echo an env var that is part of our context"
          command: |
            echo $MY_ENV_VAR
```

Creating a context allows you to share environment variables across multiple projects, and control who has access. For more information, see the [Contexts documentation]({{ site.baseurl }}/2.0/contexts/).

## Setting an Environment Variable in a Project

> 1. In the CircleCI application, go to your project's settings by clicking the gear icon on the Pipelines page, por the three dots on other pages in the application.

Reviewer: ```suggestion
1. In the CircleCI application, go to your project's settings by clicking the gear icon on the Pipelines page, or the three dots on other pages in the application.
```
